### ISSUE TITLE
Breaking on managed entry point

### ISSUE THREAD

**Issue by bendono** [dotnet/coreclr#3684]
When debugging a managed full dotnet application, it is common to break once clr is loaded.
This can be done with "sxe ld clr".
Then managed breakpionts !bpmd can be set as needed.

However, things are quite different in CoreCLR.
If you break on coreclr load (sxe ld coreclr), SOS can be loaded but the managed assembly has not yet been loaded yet.
So next I place a break on the managed assembly (sxe ld HelloWorld).
Now I place a managed breakpoint on the managed entry point:

0:000> !sos.bpmd HelloWorld Program.Main
Adding pending breakpoints...

Lets check name2ee output:
0:000> !sos.name2ee Program HelloWorld

No output at all. Probably the managed initialization is not yet complete.
But there is a pending breakpoint, so lets continue execution.

0:000> g
(fd0.3788): CLR exception - code e0434352 (first chance)
(fd0.3788): CLR exception - code e0434352 (!!! second chance !!!)
KERNELBASE!RaiseException+0x68:
00007ff9`239b1f28 488b8c24c0000000 mov     rcx,qword ptr [rsp+0C0h] ss:000000d8`2fd7cb00=000081b82481ac89

Unfortunately, it does not stop at the managed entry point, but rather continues until (an expected) Exception is thrown.

0:000> !sos.clrstack
OS Thread Id: 0x3788 (0)
        Child SP               IP Call Site
000000d82fd7d038 000000d82fd7d340 00007ff88f4728c5 Program.Error()
000000d82fd7d390 00007ff88f472821 Program.C()
000000d82fd7d3c0 00007ff88f472595 Program.B()
000000d82fd7d400 00007ff88f4707c1 Program.A()
000000d82fd7d430 00007ff88f4704b7 Program.Main(System.String[])
000000d82fd7dc88 

Also notice that the last stack frame looks a little odd...
As is the native callstack:

0:000> k
 # Child-SP          RetAddr           Call Site
00 000000d8`2fd7ca40 00007ff8`ede9d763 KERNELBASE!RaiseException+0x68
01 000000d8`2fd7cb20 00007ff8`ede96f96 CoreCLR!`RaiseTheExceptionInternalOnly'::`81'::__Body::Run+0x373 [d:\git\coreclr\src\vm\excep.cpp @ 3114]
02 000000d8`2fd7cc60 00007ff8`edf47776 CoreCLR!RaiseTheExceptionInternalOnly+0x6d6 [d:\git\coreclr\src\vm\excep.cpp @ 3114]
Unable to read dynamic function table entries
Unable to read dynamic function table entries
03 000000d8`2fd7ce40 00007ff8`8f4728c5 CoreCLR!IL_Throw+0x466 [d:\git\coreclr\src\vm\jithelpers.cpp @ 5378]
Unable to read dynamic function table entries
Unable to read dynamic function table entries
Unable to read dynamic function table entries
04 000000d8`2fd7d340 00000228`0000aff8 0x00007ff8`8f4728c5
Unable to read dynamic function table entries
05 000000d8`2fd7d348 00000228`0000b090 0x00000228`0000aff8
06 000000d8`2fd7d350 00000228`0000a2a4 0x00000228`0000b090
07 000000d8`2fd7d358 00007ff8`eef1a058 0x00000228`0000a2a4
08 000000d8`2fd7d360 00000228`0000a6e0 CoreCLR!`string'
09 000000d8`2fd7d368 00000228`0000ad80 0x00000228`0000a6e0
0a 000000d8`2fd7d370 00000228`0000aff8 0x00000228`0000ad80
0b 000000d8`2fd7d378 000000d8`2fd7d470 0x00000228`0000aff8
0c 000000d8`2fd7d380 000000d8`2fd7d3b0 0x000000d8`2fd7d470
Unable to read dynamic function table entries
Unable to read dynamic function table entries
0d 000000d8`2fd7d388 00007ff8`8f472821 0x000000d8`2fd7d3b0
Unable to read dynamic function table entries
Unable to read dynamic function table entries
Unable to read dynamic function table entries
0e 000000d8`2fd7d390 00000228`0000a2a8 0x00007ff8`8f472821
Unable to read dynamic function table entries
0f 000000d8`2fd7d398 00000000`00010000 0x00000228`0000a2a8
10 000000d8`2fd7d3a0 00000228`0000a2a4 0x10000
11 000000d8`2fd7d3a8 00007ff8`eef1a058 0x00000228`0000a2a4
12 000000d8`2fd7d3b0 000000d8`2fd7d3f0 CoreCLR!`string'
Unable to read dynamic function table entries
Unable to read dynamic function table entries
13 000000d8`2fd7d3b8 00007ff8`8f472595 0x000000d8`2fd7d3f0
Unable to read dynamic function table entries
Unable to read dynamic function table entries
Unable to read dynamic function table entries
14 000000d8`2fd7d3c0 00000228`0000a2a8 0x00007ff8`8f472595
Unable to read dynamic function table entries
15 000000d8`2fd7d3c8 00000000`00010000 0x00000228`0000a2a8
16 000000d8`2fd7d3d0 00000228`0000a2a4 0x10000
17 000000d8`2fd7d3d8 00007ff8`eef1a058 0x00000228`0000a2a4
18 000000d8`2fd7d3e0 00000000`00000000 CoreCLR!`string'

I would expect to see ntdll!RtlUserThreadStart at the bottom of the native stack.
Something similar to the following:
...
CoreCLR!CorHost2::ExecuteAssembly+0x37b [d:\git\coreclr\src\vm\corhost.cpp @ 1329]
CoreRun!TryRun+0xc59 [d:\git\coreclr\src\coreclr\hosts\corerun\corerun.cpp @ 572]
CoreRun!wmain+0x14b [d:\git\coreclr\src\coreclr\hosts\corerun\corerun.cpp @ 685]
CoreRun!invoke_main+0x22 [f:\dd\vctools\crt\vcstartup\src\startup\exe_common.inl @ 89]
CoreRun!__scrt_common_main_seh+0x124 [f:\dd\vctools\crt\vcstartup\src\startup\exe_common.inl @ 264]
KERNEL32!BaseThreadInitThunk+0x22
ntdll!RtlUserThreadStart+0x34

Restart the application and try again.
With a bit of experimentation, it seems that it works if I place a native breakpoint on CoreCLR!RunMain.
0:000> bp CoreCLR!RunMain
0:000> g
0:000> .loadby sos coreclr
Module:      00007ff886d73e60
Assembly:    HelloWorld.exe
Token:       0000000006000004
MethodDesc:  00007ff886d75150
Name:        Program.Main(System.String[])
Not JITTED yet. Use !bpmd -md 00007ff886d75150 to break on run.

Things look better now.
Set the managed breakpoint.

0:000> !sos.bpmd HelloWorld Program.Main
Found 1 methods in module 00007ff886d73e60...
MethodDesc = 00007ff886d75150
Adding pending breakpoints...
0:000> g
(3ac0.2f08): CLR notification exception - code e0444143 (first chance)
JITTED HelloWorld!Program.Main(System.String[])
Setting breakpoint: bp 00007FF8871104B1 [Program.Main(System.String[])]
Breakpoint 1 hit
Unable to read dynamic function table entries
00007ff8`871104b1 90              nop
0:000> !sos.clrstack
OS Thread Id: 0x2f08 (0)
        Child SP               IP Call Site
000000c741d7d790 00007ff8871104b1 Program.Main(System.String[])
000000c741d7dfe8 

Finally, I am able to break on the managed entry point.
However, this required that I have symbols for coreclr, which might not always be the case.

In summary:
1) Is there a way to break when CoreCLR initialization is complete? Without symbols?
2) Is there an easier way to break on the managed entry point without need for coreclr symbols?
3) The last frame of managed stacks does not look correct. Is this expected?
4) Once some managed code is executed, the native stacktrace looks extremely odd. Is this expected?

**Comment by bendono** [comment]
For reference, the HelloWorld program is slightly modified for experimenting with setting breakpoints.

`using System;

public class Foo
{
	public Foo()
	{
		Console.WriteLine("[Foo] Constructor");
	}
	
	public void Set(object o)
	{
		Console.WriteLine("[Foo] Set");
		
		if (o != null)
		{
			InternalSet(o);
		}
	}
	
	private void InternalSet(object o)
	{
		Console.WriteLine("[Foo] InternalSet: {0}", o);
	}
}

public class Program
{
	private static Foo foo = new Foo();
	
    public static void Main (string[] args)
    {
		A();
		foo.Set(123);
    }
	
	static void A()
	{
		Console.WriteLine("A");
		B();
	}
	
	static void B()
	{
		Console.WriteLine("B");
		foo.Set("abc");
		C();
	}
	
	static void C()
	{
		Console.WriteLine("C");
		Error();
	}
	
	static void Error()
	{
		foo.Set("Error!");
		throw new Exception();
	}
}
`

**Comment by mikem8361** [comment]
1) Setting these two environment variables will cause coreclr to break in early initialization without symbols:

export COMPlus_BreakOnEELoad=2
export COMPlus_DebugBreakOnAssert=1

2-4) I need to investigate some more.

**Comment by mikem8361** [comment]
I've looked at your problems again and found:

1) Is there a way to break when CoreCLR initialization is complete? Without symbols?
That "sxe ld CoreClr" allows me to set a managed breakpoint. !clrstack or !threads doesn't work after it stops, but "!bpmd HelloWorld Program.Main" worked and stopped on my Main.

 2) Is there an easier way to break on the managed entry point without need for coreclr symbols?
See above.

 3) The last frame of managed stacks does not look correct. Is this expected?
There is a linefeed problem and maybe more covered under issue #4432.

 4) Once some managed code is executed, the native stacktrace looks extremely odd. Is this expected?
That is covered under issue #4402.  If you execute this windbg command it will fix the stack trace:
.settings set EngineInitialization.VerifyFunctionTableCallbacks=false

**Comment by mikem8361** [comment]
I think all your issues are either answered or fixed.  Reopen this issue if not the case.
